qa/L0_python_api/test_kserve.py

```diff
@@ -357,6 +418,31 @@ def test_metrics_update(self, frontend, url):
         utils.teardown_service(metrics_service)
         utils.teardown_server(server)
 
+    @pytest.mark.parametrize(
+        "frontend, client_type, url, key_prefix",
+        [HTTP_ARGS + ("",), GRPC_ARGS + ("triton-grpc-protocol-",)],
+    )
+    def test_restricted_features(self, frontend, client_type, url, key_prefix):
+        server = utils.setup_server()
+        # Specifying Restricted Features
+        options = frontend.Options(restricted_features=RESTRICTED_FEATURE_ARG)
+        service = utils.setup_service(server, frontend, options=options)
+
+        # Valid headers sent with inference request
+        headers = {key_prefix + "infer-key": "infer-value"}
+        assert utils.send_and_test_inference_identity(client_type, url, headers)
+
+        # Invalid headers sent with inference request
+        headers = {key_prefix + "fake-key": "fake-value"}
+        with pytest.raises(
+            InferenceServerException,
+            match=f"expecting header '{key_prefix}infer-key'",
+        ):
+            utils.send_and_test_inference_identity(client_type, url, headers)
+
+        utils.teardown_service(service)
+        utils.teardown_server(server)
+
```

Comment:
test for:
1. multiple feature groups
2. overwrite feature groups
3. remove feature groups
4. duplicate feature in different groups

Reply:
1. Added a test for multiple feature groups (`TestRestrictedFeatures::test_multiple_groups()`)
2. Overwriting was complicating the logic for validating types and tracking duplicates, so I decided to set `@dataclass(frozen=True)` for `FeatureGroup` to prevent any post-initialization setting of attributes. So, currently if someone wants to overwrite a certain attribute of a `FeatureGroup` instance, they will have to remove the `FeatureGroup` instance from the `RestrictedFeatures` object and make a new `FeatureGroup` object to reflect their choices.
3. Tests for this have been added to `TestRestrictedFeatureOptions`
4. Tests for this have been added to `TestRestrictedFeatureOptions`

Reply:
This should be possible with `update_feature_group(Feature, key, value)` now.